File: api/mcp-scan.js
// api/mcp-scan.js — Cyber-Guardian Dashboard API (Vercel Node.js)
// Routes: ?action=stats | servers | threats | history | trigger

const { createClient } = require('@supabase/supabase-js');

const SUPABASE_URL = process.env.SUPABASE_URL;
const SUPABASE_KEY = process.env.SUPABASE_SERVICE_KEY;
const API_SECRET   = process.env.SCANNER_API_SECRET || '';
const ALLOWED_ORIGINS = (process.env.ALLOWED_ORIGINS || ('https://cyberguardianscan.com,https://cyber-guardian-mu.vercel.app' + (process.env.VERCEL_ENV === 'production' ? '' : ',http://localhost:3000,http://localhost:5173')))
  .split(',').map(s => s.trim()).filter(Boolean);

function getSupabase() {
  return createClient(SUPABASE_URL, SUPABASE_KEY);
}

function getOrigin(req) {
  return req.headers.origin || '';
}

function isAllowedOrigin(origin) {
  if (!origin) return true;
  return ALLOWED_ORIGINS.includes(origin);
}

function setCors(req, res) {
  const origin = getOrigin(req);
  if (origin && isAllowedOrigin(origin)) {
    res.setHeader('Access-Control-Allow-Origin', origin);
    res.setHeader('Vary', 'Origin');
  }
  res.setHeader('Access-Control-Allow-Methods', 'GET, POST, OPTIONS');
  res.setHeader('Access-Control-Allow-Headers', 'Content-Type, Authorization');
  res.setHeader('Content-Type', 'application/json');
}

function rejectDisallowedOrigin(req, res) {
  if (isAllowedOrigin(getOrigin(req))) return false;
  res.status(403).json({ error: 'Origin not allowed' });
  return true;
}

function clampInt(value, fallback, min, max) {
  const parsed = parseInt(value, 10);
  if (!Number.isFinite(parsed)) return fallback;
  return Math.max(min, Math.min(max, parsed));
}

async function actionStats(sb, res) {
  const { data: runs }    = await sb.from('mcp_scan_runs').select('id,started_at,completed_at,total_scanned,total_malicious,by_source,by_risk_level,by_category').order('started_at', { ascending: false }).limit(1);
  const { data: servers } = await sb.from('mcp_servers').select('risk_level, source');
  const { data: cats }    = await sb.from('mcp_threat_category_counts').select('category,total');
  const { data: trend }   = await sb.from('mcp_scan_runs').select('started_at,total_scanned,total_malicious').order('started_at', { ascending: false }).limit(7);

  const riskCounts = {}, srcCounts = {};
  for (const r of (servers || [])) {
    riskCounts[r.risk_level] = (riskCounts[r.risk_level] || 0) + 1;
    srcCounts[r.source]      = (srcCounts[r.source] || 0) + 1;
  }

  res.status(200).json({
    latest_run:        runs?.[0] || {},
    risk_counts:       riskCounts,
    source_counts:     srcCounts,
    threat_categories: cats || [],
    scan_trend:        [...(trend || [])].reverse(),
    total_servers:     Object.values(riskCounts).reduce((a,b) => a+b, 0),
    total_threats:     (cats || []).reduce((a, r) => a + (r.total || 0), 0),
  });
}

async function actionServers(sb, params, res) {
  const page    = clampInt(params.get('page'), 1, 1, 10000);
  const perPage = clampInt(params.get('per_page'), 50, 1, 100);
  const source  = params.get('source');
  const risk    = params.get('risk_level');
  const offset  = (page - 1) * perPage;

  let query = sb.from('mcp_servers')
    .select('id,name,source,url,description,stars,language,owner,risk_score,risk_level,threat_count,scan_date,files_scanned')
    .order('risk_score', { ascending: false })
    .range(offset, offset + perPage - 1);

  if (source) query = query.eq('source', source);
  if (risk)   query = query.eq('risk_level', risk);

  const { data } = await query;
  res.status(200).json({ page, per_page: perPage, servers: data || [] });
}

async function actionThreats(sb, params, res) {
  const limit  = clampInt(params.get('limit'), 50, 1, 200);
  const source = params.get('source');
  let query = sb.from('mcp_threats_view').select('server_name,source,url,risk_level,risk_score,category,severity,title,description,file_path,line_number,scan_date').limit(limit);
  if (source) query = query.eq('source', source);
  const { data } = await query;
  res.status(200).json({ threats: data || [] });
}

async function actionHistory(sb, res) {
  const { data } = await sb.from('mcp_scan_runs').select('id,started_at,completed_at,total_scanned,total_malicious,by_source,by_risk_level,by_category').order('started_at', { ascending: false }).limit(30);
  res.status(200).json({ history: [...(data || [])].reverse() });
}

async function actionTrigger(body, res) {
  if (!API_SECRET) return res.status(403).json({ error: 'Trigger endpoint is disabled' });
  let payload = {};
  try { payload = typeof body === 'string' ? JSON.parse(body) : body; } catch { return res.status(400).json({ error: 'Invalid JSON' }); }
  if (payload.secret !== API_SECRET) return res.status(401).json({ error: 'Unauthorized' });
  res.status(200).json({ status: 'Use GitHub Actions to trigger a scan', note: 'Go to Actions tab → Daily MCP Security Scan → Run workflow' });
}

module.exports = async function handler(req, res) {
  setCors(req, res);
  if (req.method === 'OPTIONS') {
    if (rejectDisallowedOrigin(req, res)) return;
    return res.status(204).end();
  }
  if (rejectDisallowedOrigin(req, res)) return;

  if (!SUPABASE_URL || !SUPABASE_KEY) return res.status(500).json({ error: 'Supabase not configured' });

  const url    = new URL(req.url, `https://${req.headers.host}`);
  const params = url.searchParams;
  const action = params.get('action') || 'stats';
  const sb     = getSupabase();

  try {
    if (req.method === 'GET') {
      if (action === 'stats')   return await actionStats(sb, res);
      if (action === 'servers') return await actionServers(sb, params, res);
      if (action === 'threats') return await actionThreats(sb, params, res);
      if (action === 'history') return await actionHistory(sb, res);
      return res.status(400).json({ error: `Unknown action: ${action}` });
    }
    if (req.method === 'POST' && action === 'trigger') return await actionTrigger(req.body, res);
    return res.status(405).json({ error: 'Method not allowed' });
  } catch (err) {
    console.error('[mcp-scan]', err);
    return res.status(500).json({ error: 'Server error' });
  }
};

File: api/mcp-scan.py
"""
api/mcp-scan.py  —  Vercel Serverless Function (Python runtime)
================================================================
Exposes the MCP scanner results to the Cyber-Guardian dashboard.

Routes (via query param ?action=...):
  GET  ?action=stats          → latest scan summary + charts data
  GET  ?action=servers        → all scanned servers (paginated)
  GET  ?action=threats        → high/critical servers only
  GET  ?action=history        → last N scan runs
  POST ?action=trigger        → manual scan trigger (protected by API_SECRET)
"""

from http.server import BaseHTTPRequestHandler
from urllib.parse import urlparse, parse_qs
import json
import os
import subprocess
import sys

from supabase import create_client

SUPABASE_URL  = os.environ["SUPABASE_URL"]
SUPABASE_KEY  = os.environ["SUPABASE_SERVICE_KEY"]   # service role for API
API_SECRET    = os.environ.get("SCANNER_API_SECRET", "")


def get_supabase():
    return create_client(SUPABASE_URL, SUPABASE_KEY)


def cors_headers() -> dict:
    return {
        "Content-Type": "application/json",
        "Access-Control-Allow-Origin": "*",
        "Access-Control-Allow-Methods": "GET, POST, OPTIONS",
        "Access-Control-Allow-Headers": "Content-Type, Authorization",
    }


def json_response(data: dict, status: int = 200) -> dict:
    return {
        "statusCode": status,
        "headers": cors_headers(),
        "body": json.dumps(data, default=str),
    }


def error_response(msg: str, status: int = 400) -> dict:
    return json_response({"error": msg}, status)


# ──────────────────────────────────────────────
#  ACTION HANDLERS
# ──────────────────────────────────────────────

def action_stats(sb) -> dict:
    """Dashboard overview: counts, chart data, latest run info."""
    # Latest scan run
    run_res = sb.table("mcp_scan_runs").select("*").order(
        "started_at", desc=True
    ).limit(1).execute()
    latest_run = run_res.data[0] if run_res.data else {}

    # Risk level distribution (all time latest scan)
    risk_res = sb.table("mcp_servers").select(
        "risk_level"
    ).execute()
    risk_counts: dict[str, int] = {}
    for row in (risk_res.data or []):
        lvl = row["risk_level"]
        risk_counts[lvl] = risk_counts.get(lvl, 0) + 1

    # Source distribution
    src_res = sb.table("mcp_servers").select("source").execute()
    src_counts: dict[str, int] = {}
    for row in (src_res.data or []):
        s = row["source"]
        src_counts[s] = src_counts.get(s, 0) + 1

    # Top threat categories (from the view)
    cat_res = sb.table("mcp_threat_category_counts").select("*").execute()

    # Trend: last 7 scan runs
    trend_res = sb.table("mcp_scan_runs").select(
        "started_at,total_scanned,total_malicious"
    ).order("started_at", desc=True).limit(7).execute()

    return json_response({
        "latest_run":     latest_run,
        "risk_counts":    risk_counts,
        "source_counts":  src_counts,
        "threat_categories": cat_res.data or [],
        "scan_trend":     list(reversed(trend_res.data or [])),
        "total_servers":  sum(risk_counts.values()),
        "total_threats":  sum(
            row.get("total", 0) for row in (cat_res.data or [])
        ),
    })


def action_servers(sb, params: dict) -> dict:
    """Paginated list of all scanned servers."""
    page     = int(params.get("page", ["1"])[0])
    per_page = min(int(params.get("per_page", ["50"])[0]), 100)
    source   = params.get("source", [None])[0]
    risk     = params.get("risk_level", [None])[0]
    offset   = (page - 1) * per_page

    query = sb.table("mcp_servers").select(
        "id,name,source,url,description,stars,language,owner,"
        "risk_score,risk_level,threat_count,scan_date,files_scanned"
    ).order("risk_score", desc=True).range(offset, offset + per_page - 1)

    if source:
        query = query.eq("source", source)
    if risk:
        query = query.eq("risk_level", risk)

    res = query.execute()
    return json_response({
        "page":     page,
        "per_page": per_page,
        "servers":  res.data or [],
    })


def action_threats(sb, params: dict) -> dict:
    """High and critical risk servers with full threat details."""
    limit  = min(int(params.get("limit", ["50"])[0]), 200)
    source = params.get("source", [None])[0]

    query = sb.table("mcp_threats_view").select("*").limit(limit)
    if source:
        query = query.eq("source", source)

    res = query.execute()
    return json_response({"threats": res.data or []})


def action_history(sb) -> dict:
    """Last 30 scan run records for the trend chart."""
    res = sb.table("mcp_scan_runs").select("*").order(
        "started_at", desc=True
    ).limit(30).execute()
    return json_response({"history": list(reversed(res.data or []))})


def action_trigger(sb, body: str) -> dict:
    """
    Manually trigger a scan.
    POST body: {"secret": "<SCANNER_API_SECRET>"}
    Runs the scanner as a background subprocess (fire-and-forget).
    Note: Vercel functions time out at 60s — for production use a
    GitHub Actions manual dispatch instead (see the workflow file).
    """
    if not API_SECRET:
        return error_response("Trigger endpoint is disabled (no API_SECRET set)", 403)

    try:
        payload = json.loads(body or "{}")
    except json.JSONDecodeError:
        return error_response("Invalid JSON body", 400)

    if payload.get("secret") != API_SECRET:
        return error_response("Unauthorized", 401)

    # Fire-and-forget — Vercel will kill the function before it finishes,
    # but this is useful for local testing via the Vercel CLI.
    subprocess.Popen(
        [sys.executable, "mcp_scanner.py"],
        env=os.environ.copy(),
    )
    return json_response({"status": "scan triggered", "note": "runs in background"})


# ──────────────────────────────────────────────
#  VERCEL HANDLER
# ──────────────────────────────────────────────

def handler(request, context):
    """Vercel Python serverless function entry point."""
    method = request.get("method", "GET").upper()
    parsed = urlparse(request.get("url", ""))
    params = parse_qs(parsed.query)
    action = params.get("action", ["stats"])[0]

    # CORS preflight
    if method == "OPTIONS":
        return {"statusCode": 204, "headers": cors_headers(), "body": ""}

    sb = get_supabase()

    try:
        if method == "GET":
            if action == "stats":
                return action_stats(sb)
            elif action == "servers":
                return action_servers(sb, params)
            elif action == "threats":
                return action_threats(sb, params)
            elif action == "history":
                return action_history(sb)
            else:
                return error_response(f"Unknown action: {action}")

        elif method == "POST":
            if action == "trigger":
                body = request.get("body", "")
                return action_trigger(sb, body)
            else:
                return error_response("Unknown POST action")

        else:
            return error_response("Method not allowed", 405)

    except Exception as e:
        return error_response(f"Server error: {str(e)}", 500)
